File: src/internal/api/companies.go
package api

import (
	"errors"
	"net/http"
	"strings"

	"github.com/go-chi/chi/v5"

	"github.com/ranaroussi/minigun/internal/store"
)

type createCompanyReq struct {
	Name string `json:"name"`
	Slug string `json:"slug"`
}

func (s *Server) handleCreateCompany(w http.ResponseWriter, r *http.Request) {
	var req createCompanyReq
	if err := decodeJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	req.Name = strings.TrimSpace(req.Name)
	req.Slug = strings.ToLower(strings.TrimSpace(req.Slug))
	if req.Name == "" {
		writeError(w, http.StatusBadRequest, "name is required")
		return
	}
	if !slugRE.MatchString(req.Slug) {
		writeError(w, http.StatusBadRequest, "slug must be lowercase alphanumerics or hyphens, 1-64 chars")
		return
	}
	c, err := s.store.CreateCompany(r.Context(), req.Slug, req.Name)
	if err != nil {
		if errors.Is(err, store.ErrAlreadyExists) {
			writeError(w, http.StatusConflict, "company with that slug already exists")
			return
		}
		s.log.Error("create company", "err", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	writeJSON(w, http.StatusCreated, c)
}

func (s *Server) handleListCompanies(w http.ResponseWriter, r *http.Request) {
	items, err := s.store.ListCompanies(r.Context())
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	if items == nil {
		items = []store.CompanySummary{}
	}
	writeJSON(w, http.StatusOK, map[string]any{"items": items})
}

func (s *Server) handleGetCompany(w http.ResponseWriter, r *http.Request) {
	key := chi.URLParam(r, "company")
	c, err := s.store.ResolveCompany(r.Context(), key)
	if err != nil {
		if errors.Is(err, store.ErrNotFound) {
			writeError(w, http.StatusNotFound, "company not found")
			return
		}
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, c)
}

func (s *Server) handleListCompanyLists(w http.ResponseWriter, r *http.Request) {
	key := chi.URLParam(r, "company")
	c, err := s.store.ResolveCompany(r.Context(), key)
	if err != nil {
		if errors.Is(err, store.ErrNotFound) {
			writeError(w, http.StatusNotFound, "company not found")
			return
		}
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	items, err := s.store.ListsForCompany(r.Context(), c.ID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"company": c,
		"items":   items,
	})
}

File: worker/src/routes/companies.ts
import { Hono } from 'hono';
import { Env } from '../env';
import { isValidSlug } from '../lib/slug';
import {
  createCompany,
  getCompanyByID,
  listCompanies,
  listsForCompany,
  resolveCompany,
} from '../store/companies';
import { AlreadyExistsError, NotFoundError } from '../store/types';

export function mountCompanies(app: Hono<{ Bindings: Env }>) {
  app.post('/companies', async (c) => {
    const body = await c
      .req.json<{ name?: string; slug?: string; domain?: string }>()
      .catch(() => null);
    if (!body) return c.json({ error: 'invalid JSON' }, 400);
    const name = (body.name ?? '').trim();
    const slug = (body.slug ?? '').trim().toLowerCase();
    const domain = (body.domain ?? '').trim().toLowerCase();
    if (!name) return c.json({ error: 'name is required' }, 400);
    if (!isValidSlug(slug)) {
      return c.json({ error: 'slug must be lowercase alphanumerics or hyphens, 1-64 chars' }, 400);
    }
    if (!domain) {
      return c.json({ error: 'domain is required (Mailgun sending domain for this company)' }, 400);
    }
    try {
      const created = await createCompany(c.env.DB, slug, name, domain);
      return c.json(created, 201);
    } catch (err) {
      if (err instanceof AlreadyExistsError) {
        return c.json({ error: 'company with that slug already exists' }, 409);
      }
      throw err;
    }
  });

  app.get('/companies', async (c) => {
    const items = await listCompanies(c.env.DB);
    return c.json({ items });
  });

  app.get('/companies/:company', async (c) => {
    const key = c.req.param('company');
    try {
      const company = await resolveCompany(c.env.DB, key);
      return c.json(company);
    } catch (err) {
      if (err instanceof NotFoundError) return c.json({ error: 'company not found' }, 404);
      throw err;
    }
  });

  app.get('/companies/:company/lists', async (c) => {
    const key = c.req.param('company');
    try {
      const company = await resolveCompany(c.env.DB, key);
      const items = await listsForCompany(c.env.DB, company.id);
      return c.json({ company, items });
    } catch (err) {
      if (err instanceof NotFoundError) return c.json({ error: 'company not found' }, 404);
      throw err;
    }
  });
}
